db/db.go

```diff
@@ -0,0 +1,260 @@
+	UpdateEntities(query.Query, Entity) ([]Entity, error)
+	DeleteEntities(query.Query) ([]Entity, error)
+
+	// Managing connection to DB
+	Connect() error
+	Disconnect()
+}
+
+// mongo struct stores all info needed to connect and query a mongo instance.
+type mongo struct {
+	url        string
+	database   string
+	collection string
+	session    *mgo.Session
+}
+
+// NewConnector creates an instance of Connector.
+func NewConnector(url string, database string, collection string) Connector {
+	return &mongo{
+		url:        url,
+		database:   database,
+		collection: collection,
+	}
+}
+
+// Connect creates a session to db. All queries to DB will copy this session to
+// make a request (https://godoc.org/gopkg.in/mgo.v2#session.Copy).
+func (m *mongo) Connect() error {
+	session, err := mgo.Dial(m.url)
+	if err != nil {
+		return ErrConnect
+	}
+
+	m.session = session
+	return nil
+}
+
+// Disconnect must be called when finished with db connector.
+func (m *mongo) Disconnect() {
+	m.session.Close()
+}
+
+// CreateEntities inserts many Entity into DB.
+func (m *mongo) CreateEntities(entities []Entity) []string {
+	s := m.session.Copy()
+	defer s.Close()
+
+	c := m.session.DB(m.database).C(m.collection)
+
+	var objectIDs []string
+
+	for _, e := range entities {
+		objectID := bson.NewObjectId().String()
+		objectIDs = append(objectIDs, objectID)
+		// Note: this modifies entity in memory
+		e["_id"] = objectID
+
+		c.Insert(e)
+	}
+
+	return objectIDs
+}
+
+// Query mongoDB instance with a Query object, which is tranlsated to bson.M
+// object for mgo driver to understand. Returns a list of Entity objects. If
+// nothing is found, returns empty list.
+func (m *mongo) ReadEntities(q query.Query) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	var entities []Entity
+	c := s.DB(m.database).C(m.collection)
+	err := c.Find(mgoQuery).All(&entities)
+	if err != nil {
+		return nil, ErrNotFound
+	}
+
+	return entities, nil
+}
+
+// UpdateEntities does a db.collection.findAndModify() operation
+// (https://docs.mongodb.com/manual/reference/method/db.collection.findAndModify/)
+// on all docs returned from query. Returns diff of entities (old values for
+// fields changed, in addition to including "_id" field/value) or an error, if
+// one occured.
+func (m *mongo) UpdateEntities(q query.Query, u Entity) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	c := s.DB(m.database).C(m.collection)
+
+	// List of IDs for docs to update
+	ids, err := idsForQuery(c, mgoQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// Change to make
+	change := mgo.Change{Update: bson.M{"$set": u}}
+
+	// Query for each document and apply update
+	var diffs []Entity
+	for _, id := range ids {
+		var diff Entity
+		_, err := c.Find(bson.M{"_id": id}).Select(selectMap(u)).Apply(change, &diff)
+		if err != nil {
+			return nil, ErrUpdateEntity
```

Comment:
You're correct. This is a case where where we're making many calls to the DB rather than just one, which makes the return value and error not mutually exclusive.

I've updated it to return any entities we actually updated and an error if there is one. along with a comment at the top of the method.